state/unit.go

```diff
@@ -405,22 +405,35 @@ func (u *Unit) destroyOps() ([]txn.Op, error) {
 
 	// See if the unit agent has started running.
 	// If so then we can't set directly to dead.
+	isAssigned := u.doc.MachineId != ""
 	agentStatusDocId := u.globalAgentKey()
 	agentStatusInfo, agentErr := getStatus(u.st, agentStatusDocId, "agent")
 	if errors.IsNotFound(agentErr) {
 		return nil, errAlreadyDying
 	} else if agentErr != nil {
 		return nil, errors.Trace(agentErr)
 	}
-	if agentStatusInfo.Status != status.Allocating {
+	if isAssigned && agentStatusInfo.Status != status.Allocating {
 		return setDyingOps, nil
 	}
+	if agentStatusInfo.Status != status.Error && agentStatusInfo.Status != status.Allocating {
+		return nil, errors.Errorf("unexpected unit state - unit with status %v is not assigned to a machine", agentStatusInfo.Status)
+	}
 
-	ops := []txn.Op{{
-		C:      statusesC,
-		Id:     u.st.docID(agentStatusDocId),
-		Assert: bson.D{{"status", status.Allocating}},
-	}, minUnitsOp}
+	// TODO(wallyworld) - we really want to assert on status==Error && machineid==""
+	// but we can't do that across collections. For now, if a unit cannot be assigned,
+	// it will stay in that state and status and machineid will not change.
+	statusOp := txn.Op{
+		C:  statusesC,
+		Id: u.st.docID(agentStatusDocId),
+	}
+	if isAssigned {
+		statusOp.Assert = bson.D{{"status", status.Allocating}}
+	} else {
+		statusOp.Assert = bson.D{{"$or", []bson.D{{{"status", status.Allocating}}, {{"status", status.Error}}}}}
```

Comment:
I don't think this is good enough. A unit could be concurrently assigned to a machine when it's in error state (esp. once the assigner retries automatically).

We already do multi-collection assertions (statusOp, minUnitsOp, and isAliveDoc all assert on different collections). Please check machineid is empty when status is error.